app/controllers/add-preprint.js

```diff
@@ -1,87 +1,207 @@
+            this.get('panelActions').open(this.get(`_names.${this.get('_names').indexOf(name) + 1}`));
         },
-        error(ignore, dropzone, file, errorMessage) {
-            console.error(errorMessage);
-            this.get('toast').error(`Error uploading file:${errorMessage}`);
+        changeState(newState) {
+            this.set('state', newState);
         },
-        buildUrl() {
-            return this.get('_url');
+        changeUploadState(newState) {
+            this.set('uploadState', newState);
         },
-        uploadNewPreprintToNewProject(title, abstract, authors, subject, tags, journal, doi) {
-            //Create new public project
-            let node = this.get('store').createRecord('node', {
-                title: title,
+        createProject(resolve) {
+            const node = this.get('store').createRecord('node', {
+                title: this.get('nodeTitle'),
                 category: 'project',
-                description: abstract || null,
                 public: true
             });
-
-            this.set('preprint', {
-                title: title,
-                abstract: abstract,
-                authors: authors,
-                subject: subject,
-                tags: tags,
-                journal: journal,
-                doi: doi
-            });
-
-            // TODO: Add logic for if you're not uploading a file
-            // Upload a new file to the new project
             node.save().then(() => {
-                this.set('_url', `https://test-files.osf.io/file?path=/${this.get('latestFileName')}&nid=${node.id}&provider=osfstorage`);
-                this.set('openModal', false);
-                this.get('resolve')();
-            }).catch(error => {
-                console.error(error);
-                this.get('toast').error('Could not create preprint project!');
+                this.set('preprintNode', node);
+                this._refreshNodes();
+                this.send('startUpload', resolve);
             });
-            //TODO: eventually make a call to set an OSF file as a preprint (will probably need a flag for such)
+        },
+        deleteProject(resolve) {
+            if (this.get('preprintNode')) {
+                this.get('preprintNode').destroyRecord().then(() => {
+                    this.get('toast').info('Project deleted');
+                });
+                this.set('preprintNode', null);
+                // TODO: reset dropzone, since uploaded file has no project
+            }
+            resolve();
+        },
+        preUpload(ignore, dropzone, file) {
+            this.set('uploadFile', file);
+            return new Ember.RSVP.Promise(resolve => this.set('resolve', resolve));
+        },
+        startUpload(resolve) {
+            this.set('_url', `${this.get('preprintNode.files').findBy('name', 'osfstorage').get('links.upload')}?kind=file&name=${this.get('uploadFile.name')}`);
+            this.get('resolve')();
```

Comment:
Out of curiosity: once the `resolve` method of the promise here is called, it would seem like it has done its work. Why are we still stashing it at all?

(This is one argument against stashing callbacks in some intermediate location, instead of being more explicit about the flow of data: it's easy to forget to clean up after ourselves. The solution is to change the data flow, not to blank out this field)